### PR TITLE
vip-manager addition

### PR DESCRIPTION
**Pull request by blogh**
Proposal to add a vip-manager provision.

### CODE REVIEW

File: provision/vipmanager.bash [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+done
+
+if rpm -q vip-manager &>/dev/null ; then
+    echo "vipmanager was already installed"
+else
+    echo "Installing vipmanager"
+    rpm --import "/etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7"
+    yum install -y --quiet wget
+
+    VIPMANAGER_RPM="vip-manager_${VIP_MANAGER_VERSION}-1_amd64.rpm"
```
Comment (ioguix) [inline provision/vipmanager.bash]: This should help getting the packag name without having to mess with version all over the place:

~~~
wget -qO- 'https://api.github.com/repos/cybertec-postgresql/vip-manager/releases/latest'|jq '.assets | .[] | select(.content_type == "application/x-rpm") | .browser_download_url'
~~~

File: Vagrantfile [hunk 2]
```diff
@@ -3,13 +3,15 @@ require 'ipaddr'
 
 ENV["LC_ALL"] = 'en_US.utf8'
 
+vmver         = '0.6'
```
Comment (ioguix) [inline Vagrantfile]: probably useless, see other comment

File: provision/vipmanager.bash [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        v) VIP_MANAGER_VERSION="${OPTARG}" ;;
+        e) ETCD[${#ETCD[@]}]=$OPTARG ;;
+    esac
+done
+
+if rpm -q vip-manager &>/dev/null ; then
+    echo "vipmanager was already installed"
+else
+    echo "Installing vipmanager"
+    rpm --import "/etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7"
```
Comment (ioguix) [inline provision/vipmanager.bash]: This is useless. Eg.: 
https://github.com/ClusterLabs/PAF/blob/master/extra/vagrant/3nodes-vip/provision/system.bash#L35

File: Vagrantfile [hunk 4]
```diff
@@ -52,29 +54,41 @@ Vagrant.configure(2) do |config|
...
     end
 
-    # system setup for patroni ndoes
+    # system setup for patroni nodes
     (patroni_nodes).each do |node|
         config.vm.define node do |conf|
-            args = [ '-n', node, '-v', pgver ];
+            args_pa = [ '-n', node, '-v', pgver ];
+            args_vm = [ '-n', node, '-v', vmver, '-i', vip_ip.to_s ];
+
```
Comment (ioguix) [inline Vagrantfile:59]: no need for two distinct array. Just put common args in the array, then concat additionnal args later when needed. See bellow for an example

File: Vagrantfile [hunk 5]
```diff
@@ -52,29 +54,41 @@ Vagrant.configure(2) do |config|
...
-                args.push("-e", "#{e}=#{nodes_ips[e]}")
+                args_pa.push("-e", "#{e}=#{nodes_ips[e]}")
+                args_vm.push("-e", "#{e}=#{nodes_ips[e]}")
             end
-            
+
             conf.vm.network 'private_network', ip: nodes_ips[node]
             conf.vm.provision 'patroni-setup', type: 'shell', path: 'provision/patroni.bash',
-                args: args,
+                args: args_pa,
```
Comment (ioguix) [inline Vagrantfile]: could be:

~~~
            args: args + [ '-v', pgver ];
~~~

**Review by ioguix — CHANGES_REQUESTED** [review]

File: provision/vipmanager.bash (hunk 1 above)
Comment (blogh) [reply provision/vipmanager.bash]: done that way.

File: Vagrantfile (hunk 2 above)
Comment (blogh) [reply Vagrantfile]: removed

File: provision/vipmanager.bash (hunk 3 above)
Comment (blogh) [reply provision/vipmanager.bash]: removed

File: Vagrantfile (hunk 4 above)
Comment (blogh) [reply Vagrantfile:59]: I thought it was more readable.
Done as suggested.

File: Vagrantfile (hunk 5 above)
Comment (blogh) [reply Vagrantfile]: done

**Comment by blogh** [comment]
Thx for the help & remarks.
I made an attempt at the requested changes in the last commit.

File: Vagrantfile (hunk 2 above)
Comment (ioguix) [reply Vagrantfile]: It's still there :)

**Review by ioguix — CHANGES_REQUESTED** [review]

File: Vagrantfile (hunk 2 above)
Comment (blogh) [reply Vagrantfile]: arg, done

File: Vagrantfile [hunk 6]
```diff
@@ -10,6 +10,7 @@ patroni_nodes = 'p1', 'p2'       # you must have 2+ patroni nodes
 
 Vagrant.configure(2) do |config|
 
+    vip_ip     = IPAddr.new(start_ip)
     next_ip    = IPAddr.new(start_ip).succ
     host_ip    = (IPAddr.new(start_ip) & "255.255.255.0").succ.to_s
```
Comment (blogh) [inline Vagrantfile:15]: I don't quite understand the use of this line ?

Comment (blogh) [inline Vagrantfile]: host_ip is not used afterwards. is it a leftover from the PAF vagrant file ?

File: provision/vipmanager.bash [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+done
+
+
+echo "Installing packages"
+yum install --nogpgcheck --quiet -y -e 0 epel-release
+yum install --nogpgcheck --quiet -y -e 0 jq wget
+VIP_RPM=$(wget -qO- 'https://api.github.com/repos/cybertec-postgresql/vip-manager/releases/latest'| \
+          jq -r '.assets | .[] | select(.content_type == "application/x-rpm") | .browser_download_url')
+wget -q "$VIP_RPM"
+yum install --nogpgcheck --quiet -y -e 0 $(basename "$VIP_RPM")
```
Comment (ioguix) [inline provision/vipmanager.bash]: Is it possible to use directly `yum install --nogpgcheck --quiet -y -e 0 "$VIP_RPM"` ?

File: Vagrantfile (hunk 6 above)
Comment (ioguix) [reply Vagrantfile]: indeed. Can you add a specific commit in this branch to clean it please?

**Review by ioguix — CHANGES_REQUESTED** [review]

File: Vagrantfile (hunk 6 above)
Comment (blogh) [reply Vagrantfile]: will do

File: provision/vipmanager.bash (hunk 7 above)
Comment (blogh) [reply provision/vipmanager.bash]: good catch I tried it before adding "-r" to "jq" and assumed it couldn't work.
But now it's working.

**Review by ioguix — APPROVED** [review]

**Comment by blogh** [comment]
Is there a reason we do the following in "provision/patroni.bash" :
```
# hostname and /etc/hosts setup                                                 
hostnamectl set-hostname "${NODENAME}" 
```
Instead of using : `config.vm.hostname` in the vagrant file ?

https://www.vagrantup.com/docs/vagrantfile/machine_settings.html
> config.vm.hostname (string) - The hostname the machine should have. Defaults to nil. If nil, Vagrant will not manage the hostname. If set to a string, the hostname will be set on boot. If set, Vagrant will update /etc/hosts on the guest with the configured hostname.

**Comment by ioguix** [comment]
> Instead of using : config.vm.hostname in the vagrant file ?

I don't remember why I didn't use it. It has probably been inherited from PAF vagrant files. Maybe I had some compatibility issues between vagrant and libvirtd or with some OS. Or maybe I wasn't able to use it correctly... I honestly don't remember.

Using `config.vm.hostname` would be an enhancement, it worth testing it with main distros, then open a new PR dedicated to it.

**Pull request merged**